# Homework #5

- Date: 2024.04.21
- Due Date: 2024.04.28

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

## Pandas DataFrame

- 판다스(Pandas)는 파이썬에서 데이터 조작과 분석을 위한 라이브러리
    - 주로 테이블 형태의 데이터를 다루는 데 사용(판다스의 핵심 데이터 구조는 DataFrame)
    - DataFrame은 행과 열이 있는 테이블 데이터 구조를 제공
    - 각 열은 서로 다른 유형의 데이터를 포함할 수 있으며, 각 행은 각기 다른 관측치를 나타냄

- 판다스의 DataFrame을 생성하고 조작하는 방법은 매우 다양
    - 일반적으로 CSV 파일, 엑셀 파일, SQL 데이터베이스 등 다양한 소스에서 데이터를 로드하여 DataFrame 활용

In [2]:
# 딕셔너리로부터 DataFrame 생성
data = {
    '이름': ['Alice', 'Bob', 'Charlie', 'David'],
    '나이': [25, 30, 35, 40],
    '성별': ['여', '남', '남', '여']
}

df = pd.DataFrame(data)

print(df)

        이름  나이 성별
0    Alice  25  여
1      Bob  30  남
2  Charlie  35  남
3    David  40  여


In [3]:
# '이름' 열 선택

In [4]:
# 특정 조건을 만족하는 행 제거


In [5]:
data = {
    '이름': ['Alice', 'Bob', 'Charlie', 'David'],
    '나이': [25, 30, 35, 40],
    '성별': ['여', '남', '남', '여']
}

df = pd.DataFrame(data)

# 제거할 행의 인덱스를 제외한 DataFrame 생성


- pd.concat() 함수를 사용하여 DataFrame을 연결할 때 ignore_index=True를 설정하면 인덱스가 재설정되어 새로운 DataFrame이 생성됩니다. 이를 이용하여 행을 추가할 수 있습니다. 또한, drop() 메서드를 사용하여 특정 인덱스의 행을 제거할 수 있습니다.

In [6]:
# 기존 DataFrame 생성
data1 = {
    '이름': ['Alice', 'Bob', 'Charlie'],
    '나이': [25, 30, 35],
    '성별': ['여', '남', '남']
}
df1 = pd.DataFrame(data1)

# 추가할 행 데이터
data2 = {
    '이름': ['David'],
    '나이': [40],
    '성별': ['남']
}
df2 = pd.DataFrame(data2)

# 두 DataFrame을 연결하여 행 추가


- numpy를 이용한 randaom 행렬 생성

In [7]:
# 랜덤한 데이터로 DataFrame 생성


In [8]:
# 평균과 표준편차 지정
mu, sigma = 0, 0.1  # 평균(mu)이 0이고 표준편차(sigma)가 0.1인 가우시안 분포

# 난수 생성


Question 1. Gaussian 난수 생성을 이용하여, 임의의 100명의 학생의 국어, 영어, 수학 과목 성적표를 만드시오.
- data[0] = 학번
- data[1] = 국어성적 (0~100점 사이, 평균=73, 표준편차=12.3)
- data[2] = 영어성적 (0~100점 사이, 평균=67, 표준편차=25.8)
- data[3] = 수학성적 (0~100점 사이, 평균=56, 표준편차=32.7)

* 단, 0점 이하의 값이 나오는 경우 0점으로, 100점 이상의 값이 나오는 경우 100점으로 처리

Question 2. 5번째 열(mean)을 추가하여, 학생별 평균값을 갖는 데이터를 만들어주세요

Question 3. 6번째 열(rank)을 추가하여, 등수를 구해주세요.

- stock.csv로부터 데이터 불러오기
    - index: Ticker Symbol for Indexes
    - Date: Date of Observation
    - Open: Opening Price
    - High: Highest price during trading day
    - Low: Lowest price during trading day
    - Close: Close price adjusted for splits
    - Adj Close: Adjusted close price adjusted for both dividends and splits
    - Volume: Number of shares traded during trading day

In [9]:
data = pd.read_csv("data/stock.csv")

In [10]:
info = pd.read_csv("data/stockInfo.csv")

In [11]:
info

Region                          Exchange      Index Currency
0   United States           New York Stock Exchange        NYA      USD
1   United States                            NASDAQ       IXIC      USD
2       Hong Kong          Hong Kong Stock Exchange        HSI      HKD
3           China           Shanghai Stock Exchange  000001.SS      CNY
4           Japan              Tokyo Stock Exchange       N225      JPY
5          Europe                          Euronext       N100      EUR
6           China           Shenzhen Stock Exchange  399001.SZ      CNY
7          Canada            Toronto Stock Exchange     GSPTSE      CAD
8           India  National Stock Exchange of India       NSEI      INR
9         Germany          Frankfurt Stock Exchange      GDAXI      EUR
10          Korea                    Korea Exchange       KS11      KRW
11    Switzerland                SIX Swiss Exchange       SSMI      CHF
12         Taiwan             Taiwan Stock Exchange       TWII      TWD
13   South Africa       Johannesburg Stock Exchange    J203.JO      ZAR

Stock 'Index'
- ['NYA', 'IXIC', 'HSI', '000001.SS', 'GSPTSE', '399001.SZ', 'NSEI','GDAXI', 'KS11', 'SSMI', 'TWII', 'J203.JO', 'N225', 'N100']

In [12]:
data_NYA = data[data['Index']=="NYA"]
data_IXIC = data[data['Index']=="IXIC"]
data_HSI = data[data['Index']=="HSI"]

이동평균(Moving Average) =  데이터의 이동하는 평균값을 계산하는데 사용되는 방법(주어진 기간의 데이터를 사용하여 해당 기간 동안의 평균을 계산하는 방법)

- 여기서는 pandas의 rolling() 메서드를 사용하여 이동평균을 계산하는 방법

Homework

- 위의 그림을 rolling.mean() 메소드를 사용하지 않고, 직접 구현해보기